engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java

```diff
@@ -64,19 +75,68 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
               value.getProcessInstanceKey(), targetProcessDefinitionKey));
     }
 
+    // avoid stackoverflow using a queue to iterate over the descendants instead of recursion
+    final var elementInstances = new ArrayDeque<>(List.of(processInstance));
```

Comment:
❓ Why did we prefer `ArrayDeque` over `LinkedList` here? Using a LinkedList can provide performance gain over array based queue since it doesn't need resizing of an internal array. Maybe I am missing something here still. 🤔

Reply:
👍 Interesting question

I picked it because I thought it was the defacto choice as a Java queue. 

Regarding performance: According to the [`ArrayDeque`'s JavaDoc](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/ArrayDeque.html) it is likely faster than `LinkedList` when used a queue. I haven't measured differences though. To be honest, I doubt that this would be the cause of a performance issue when there are many child element instances. Instead, I think we'd see the biggest hit from RocksDB read and writes + network IO:
>This class is likely to be faster than Stack when used as a stack, and faster than LinkedList when used as a queue.

Reply:
Hmm. I am not sure which usage they exactly referring to. E.g. they refer to using it as dequeue where you insert and delete **from both ends**. But I expect using LinkedList for only as a FIFO queue to be faster because there is not need to resizing.

I agree, I also do not expect hundred thousands child instances, so the in memory performance impact can be quite low if you compare it with I/O. 👍